adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java

```diff
@@ -168,11 +169,58 @@ public void run() {
 
                     final AcquireTokenSilentHandler acquireTokenSilentHandler = new AcquireTokenSilentHandler(mContext,
                             authenticationRequest, mTokenCacheAccessor);
-                    final AuthenticationResult authResult
+                    final AuthenticationResult result
                             = acquireTokenSilentHandler.acquireTokenWithRefreshToken(refreshToken);
+
+                    final String correlationId = String.format(" CorrelationId: %s", authenticationRequest.getCorrelationId().toString());
+
+                    if (result == null) {
+                        Logger.e(TAG + methodName, "Returned result with exchanging refresh token for access token is null" + correlationId, "",
+                                ADALError.AUTH_REFRESH_FAILED);
+                        throw new AuthenticationException(
+                                ADALError.AUTH_REFRESH_FAILED, correlationId);
+                    }
+
+                    if (!StringExtensions.isNullOrBlank(result.getErrorCode())) {
+                        Logger.e(TAG + methodName, " ErrorCode:" + result.getErrorCode(), " ErrorDescription:" + result.getErrorDescription(), ADALError.AUTH_REFRESH_FAILED);
+                        throw new AuthenticationException(ADALError.AUTH_REFRESH_FAILED,
+                                " ErrorCode:" + result.getErrorCode());
+                    }
+
+                    if (StringExtensions.isNullOrBlank(result.getAccessToken())) {
+                        Logger.e(TAG + methodName, "Access Token not returned from server", "", ADALError.AUTH_FAILED_NO_TOKEN);
+                        throw new AuthenticationException(ADALError.AUTH_FAILED_NO_TOKEN,
+                                " Access Token not returned from server ");
+                    }
+
+                    final String rawIdToken = result.getIdToken();
+
+                    if (!TextUtils.isEmpty(rawIdToken)) {
```

Comment:
A comment here about whether or not this case where we didn't get an IdToken is expected, unexpected, valid, or invalid might be useful

Reply:
Addressed